src/raft.rs

```diff
@@ -888,12 +1002,106 @@ impl<T: Storage> Raft<T> {
-            .for_each(|(id, pr)| core.send_append(*id, pr, msgs));
+            .for_each(|(id, pr)| {
+                let mut tmp_msgs = Vec::default();
+                // Let messages be pushed into tmp_vec firstly.
+                core.send_append(*id, pr, &mut tmp_msgs);
+                for msg in tmp_msgs {
+                    // Filter out messages that need to be forwarded into msg_group.
+                    // Other messages are sent directly.
+                    if pr.broadcast_group_id == leader_group_id
+                        || msg.get_msg_type() != MessageType::MsgAppend
+                        || !pr.is_replicating()
+                    {
+                        msgs.push(msg);
+                    } else {
+                        let peer_id = msg.to;
+                        core.msg_group
+                            .entry(pr.broadcast_group_id)
+                            .or_default()
+                            // The agent must be a voter and active recently.
+                            .push((msg, pr.recent_active && conf.voters().contains(peer_id)));
+                    }
+                }
+            });
+
+        // Merge messages in the same broadcast group and send them.
+        for (_, mut group) in core.msg_group.drain() {
+            let mut need_merge = group.len() > 1;
+            let mut agent_msg_idx = None;
+            if need_merge {
+                // If found an appropriate agent, return the index of agent's message. Otherwise, return None.
+                agent_msg_idx = group
+                    .iter()
+                    .position(|(_, is_agent_candidate)| *is_agent_candidate);
+                need_merge = agent_msg_idx.is_some();
+            }
+            // Do not need to merge if group size is less than two. Or there is no appropriate agent.
+            if !need_merge {
+                msgs.append(&mut group.into_iter().map(|(msg, _)| msg).collect());
+                return;
+            }
+
+            // Record forward information
+            let mut forwards: Vec<Forward> = Vec::default();
+            for (idx, (msg, _)) in group.iter().enumerate() {
+                // MsgAppend sent to other peers in this group only reserve basic forward information.
+                if idx != agent_msg_idx.unwrap() {
+                    let forward = Forward {
+                        to: msg.to,
+                        log_term: msg.log_term,
+                        index: msg.index,
+                        ..Default::default()
+                    };
+                    forwards.push(forward);
+                }
+            }
+            // Attach forward information to MsgGroupbroadcast and send it.
+            let mut agent_msg = group.swap_remove(agent_msg_idx.unwrap()).0;
+            agent_msg.set_msg_type(MessageType::MsgGroupBroadcast);
+            agent_msg.set_forwards(forwards.into());
+            msgs.push(agent_msg);
```

Comment:
no need to send other msgs for agent?

Reply:
Other messages have been pushed to `self.msgs` when generated.

Reply:
Oh, I was thought there would be multiple `MsgAppend` for one peer